db/db_impl/db_impl_open.cc

```diff
@@ -1158,266 +1171,470 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
+    // DB and there is only a single thread operating on DB
+    ColumnFamilyData* cfd;
+
+    while ((cfd = flush_scheduler_.TakeNextColumnFamily()) != nullptr) {
+      cfd->UnrefAndTryDelete();
+      // If this asserts, it means that InsertInto failed in
+      // filtering updates to already-flushed column families
+      assert(cfd->GetLogNumber() <= wal_number);
+      (void)wal_number;
+      auto iter = version_edits->find(cfd->GetID());
+      assert(iter != version_edits->end());
+      VersionEdit* edit = &iter->second;
+      status = WriteLevel0TableForRecovery(job_id, cfd, cfd->mem(), edit);
+      if (!status.ok()) {
+        // Reflect errors immediately so that conditions like full
+        // file-systems cause the DB::Open() to fail.
         return status;
       }
+      *flushed = true;
+
+      cfd->CreateNewMemtable(*cfd->GetLatestMutableCFOptions(),
+                             *next_sequence - 1);
     }
+  }
+  return status;
+}
+
+Status DBImpl::HandleNonOkStatusOrOldLogRecord(
+    uint64_t wal_number, SequenceNumber const* const next_sequence,
+    Status status, bool* old_log_record, bool* stop_replay_for_corruption,
+    uint64_t* corrupted_wal_number, bool* corrupted_wal_found) {
+  assert(!status.ok() || *old_log_record);
+
+  assert(next_sequence);
+  assert(old_log_record);
+  assert(stop_replay_for_corruption);
+  assert(corrupted_wal_number);
+
+  if (status.IsNotSupported()) {
+    // We should not treat NotSupported as corruption. It is rather a clear
+    // sign that we are processing a WAL that is produced by an incompatible
+    // version of the code.
+    return status;
+  }
 
+  if (immutable_db_options_.wal_recovery_mode ==
+      WALRecoveryMode::kSkipAnyCorruptedRecords) {
+    // We should ignore all errors unconditionally
+    return Status::OK();
+  } else if (immutable_db_options_.wal_recovery_mode ==
+             WALRecoveryMode::kPointInTimeRecovery) {
+    if (status.IsIOError()) {
+      ROCKS_LOG_ERROR(immutable_db_options_.info_log,
+                      "IOError during point-in-time reading log #%" PRIu64
+                      " seq #%" PRIu64
+                      ". %s. This likely mean loss of synced WAL, "
+                      "thus recovery fails.",
+                      wal_number, *next_sequence, status.ToString().c_str());
+      return status;
+    }
+    // We should ignore the error but not continue replaying
+    *old_log_record = false;
+    *stop_replay_for_corruption = true;
+    *corrupted_wal_number = wal_number;
+    if (corrupted_wal_found != nullptr) {
+      *corrupted_wal_found = true;
+    }
+    ROCKS_LOG_INFO(immutable_db_options_.info_log,
+                   "Point in time recovered to log #%" PRIu64 " seq #%" PRIu64,
+                   wal_number, *next_sequence);
+    return Status::OK();
+  } else {
+    assert(immutable_db_options_.wal_recovery_mode ==
+               WALRecoveryMode::kTolerateCorruptedTailRecords ||
```

Comment:
Not related to this PR, do you happen to know why `kTolerateCorruptedTailRecords` doesn't care about if this non-OK status is for the tail record?

Reply:
I guess it wants a very limited capacity to travel back in time - that is right before the crash happened https://github.com/facebook/rocksdb/blob/2ff9f95f59ae4c941f7148a1ab549e33488a6d47/include/rocksdb/options.h#L408-L410

But I think it might have ended up allowing more unintended cases.